nx/lib/nx/binary_backend.ex

```diff
@@ -1038,6 +1038,157 @@ defmodule Nx.BinaryBackend do
+      end
+      |> Nx.as_type(type)
+      |> to_binary()
+
+    shape_h = {k, k}
+
+    {q_bin, r_bin} =
+      for i <- 0..(n - 1), reduce: {nil, r} do
+        {q, r} ->
+          # a = r[[i..(k - 1), i]]
+          {a, num_rows_a, _, _} =
+            for <<x::bitstring-size(num_bits) <- r>>, reduce: {<<>>, 0, 0, 0} do
+              {data, num_rows_a_prev, row, col} ->
+                {item, num_rows_a} =
+                  if row >= i and col == i do
+                    {x, num_rows_a_prev + 1}
+                  else
+                    {"", num_rows_a_prev}
+                  end
+
+                {row, col} =
+                  if col + 1 == n do
+                    {row + 1, 0}
+                  else
+                    {row, col + 1}
+                  end
+
+                {data <> item, num_rows_a, row, col}
+            end
+
+          h_bin = householder_reflector(a, num_rows_a, type, k)
+
+          # If we haven't allocated Q yet, let Q = H1
+          q =
+            if is_nil(q) do
+              padding_length = (m - k) * k
+
+              zero = number_to_binary(0, type)
+
+              h_bin <> String.duplicate(zero, padding_length)
+            else
+              dot_binary(q, shape_q, h_bin, shape_h, type)
+            end
+
+          r = dot_binary(h_bin, shape_q, r, shape_r, type)
+
+          {q, r}
+      end
+
+    q = from_binary(q_holder, q_bin)
+    r = from_binary(r_holder, r_bin)
+
+    case mode do
+      :complete ->
+        {q, r}
+
+      :reduced ->
+        {q, r[[0..(k - 1), 0..(n - 1)]]}
+    end
+  end
+
+  defp dot_binary(b1, shape1, b2, shape2, {_, s} = type) do
+    fun = fn lhs, rhs, acc ->
+      res = binary_to_number(lhs, type) * binary_to_number(rhs, type) + acc
+      {res, res}
+    end
+
+    v1 = aggregate_axes(b1, [1], shape1, s)
+    v2 = aggregate_axes(b2, [0], shape2, s)
+
+    for b1 <- v1, b2 <- v2, into: <<>> do
+      {bin, _acc} = bin_zip_reduce_axis(b1, b2, s, s, <<>>, 0, fun)
+      scalar_to_binary(bin, type)
+    end
+  end
+
+  defp householder_reflector(a_bin, num_rows_a, {_, num_bits} = type, target_k) do
+    # We receive the target shape so the binary is already pre-generated with 1s in the
+    # main diagonal for rows < k - n
+    <<head::bitstring-size(num_bits), tail::bitstring>> = a_bin
+
+    norm_a_squared =
+      for <<x_bin::bitstring-size(num_bits) <- a_bin>>, reduce: 0 do
+        acc ->
+          x = binary_to_number(x_bin, type)
+          acc + x * x
+      end
+
+    a_0 = binary_to_number(head, type)
+
+    v_0 =
+      if a_0 > 0 do
+        a_0 + :math.sqrt(norm_a_squared)
+      else
+        a_0 - :math.sqrt(norm_a_squared)
+      end
+
+    prefix_threshold = target_k - num_rows_a
+    prefix_size = prefix_threshold * num_bits
+    v_bin = <<0::size(prefix_size)>> <> number_to_binary(v_0, type) <> tail
+
+    # dot(v, v) = norm_v_squared, which can be calculated from norm_a as: norm_v_squared = norm_a_squared - a_0^2 + v_0^2
```

Comment:
I don't know if I should document this better or it's legible enough. I'm open to suggestions.

The main trick here is that we have already needed to calculate `<a, a> = norm_a^2`. Since `v` is basically `a` with one element changed, `norm_v^2` can be calculated by subtracting the square of the original element and adding the square of the new one, as described in the comment above.

This avoids having to iterate over `v` a second time.